src/monitor/process_monitor.cpp

```diff
@@ -20,292 +20,75 @@
-        running = false;
-
-        // send attached process SIGSTOP, so the tracee enters signal-delivery-stop
-        kill(attached_pid, SIGSTOP);
-
-        return;
-    }
-
-    // Log is certainly not signal safe
-    // lo2s::Log::debug() << "sig_handler called, what should I do with signal " << signum;
-
-    // For some miraculous reason, the signal ends up at the monitored process anyhow, even if we
-    // do not explicitly forward it here.
-
-    /* restore default signal handler */
-    // signal(SIGINT, default_signal_handler);
-}
+#include <lo2s/process_map/process_map.hpp>
+#include <lo2s/process_map/thread.hpp>
 
 namespace lo2s
 {
 namespace monitor
 {
 
-void check_ptrace(enum __ptrace_request request, pid_t pid, void* addr = nullptr,
-                  void* data = nullptr)
-{
-    auto retval = ptrace(request, pid, addr, data);
-    if (retval == -1)
-    {
-        auto ex = make_system_error();
-        Log::info() << "Failed ptrace call: " << request << ", " << pid << ", " << addr << ", "
-                    << data << ": " << ex.what();
-        throw ex;
-    }
-}
-
-void check_ptrace_setoptions(pid_t pid, long options)
+ProcessMonitor::ProcessMonitor()
+    : MainMonitor()
 {
-    check_ptrace(PTRACE_SETOPTIONS, pid, NULL, (void*)options);
+    trace_.register_monitoring_tid(gettid(), "ProcessMonitor",
+            "ProcessMonitor");
 }
 
-ProcessMonitor::ProcessMonitor(pid_t child, const std::string& name, bool spawn)
-: MainMonitor(), first_child_(child), threads_(*this),
-  default_signal_handler(signal(SIGINT, sig_handler)), num_wakeups_(0)
+void ProcessMonitor::insert_first_process(pid_t pid, std::string proc_name, bool spawn)
 {
-    trace_.register_monitoring_tid(gettid(), "ProcessMonitor", "ProcessMonitor");
+    trace_.process(pid, proc_name);
 
-    if (spawn)
-    {
-        attached_pid = -1;
-    }
-    else
-    {
-        attached_pid = child;
-    }
-    running = true;
+    Process &process = process_map().get_process(pid);
+    process.info =  new ProcessInfo(pid, spawn);
 
-    trace().process(child, name);
-    threads_.insert(child, child, spawn);
+    process_map().get_thread(pid).monitor = new ThreadMonitor(pid, pid, *this, *process.info, spawn);
 }
 
-ProcessMonitor::~ProcessMonitor()
+void ProcessMonitor::insert_process(pid_t pid, std::string proc_name)
 {
-    summary().record_perf_wakeups(num_wakeups_);
-    threads_.stop();
+    trace_.process(pid, proc_name);
+    insert_thread(pid, pid);
 }
 
-void ProcessMonitor::run()
+void ProcessMonitor::insert_thread(pid_t pid, pid_t tid)
 {
-    // monitor fork/exit/... via ptrace
-    while (1)
-    {
-        /* wait for signal */
-        int status;
-        pid_t child = waitpid(-1, &status, __WALL);
-
-        num_wakeups_++;
+    Process &process = process_map().get_process(pid);
 
-        handle_signal(child, status);
+    if(process.info == nullptr)
+    {
+        process.info = new ProcessInfo(pid, false);
```

Comment:
`new`

Use std::unique_ptr<...> instead and encapsulate this whole allocation logic inside the respective class.
This is called RAII and is the fundamental principle of resource management in C++